### PR TITLE
Auto-label/order microstates maps

### PR DESCRIPTION
**Pull request by vferat**
As discussed in #82, this PR add function(s) to automatically reorder `cluster centers`:
- `optimize_order` : find the best order to maximize auto-correlation between two cluster instances.

### CODE REVIEW

**Comment by vferat** [comment]
@mscheltienne not sure where in the API to put the optimize_order function. Otherwise have a look !

**Review by mscheltienne — APPROVED** [review]
LGTM, maybe since this matches the order to another instance, we could call it `match_order` and we could add a method in the `_BaseCluster` class which takes another Cluster instance in input and matches the order directly. It would slightly shorten the user-code from:

```
from pycrostates.utils import optimize_order
from pycrostates.utils import ModKMeans

c1 = ModKMeans(...)
c2 = ModKMeans(...)
order = optimize_order(c1, c2)
c2.reorder(order)
```

to 

```

```
from pycrostates.utils import optimize_order
from pycrostates.utils import ModKMeans

c1 = ModKMeans(...)
c2 = ModKMeans(...)
c2.match_order(c1)
```

Also, if we keep the function in `utils/_clusters.py`, the public function should be added to `utils/__init__.py`?

**Comment by vferat** [comment]
I added the `'template'` argument to the `reorder_clusters`  method which uses `optimize_order`
I also moved the `utils._cluster` module to `clusters.utils` and adapted the documentation.

Let me know if you think there is still changes to do, otherwise you're free to merge.

File: pycrostates/cluster/_base.py [hunk 1]
```diff
@@ -373,19 +387,29 @@ def reorder_clusters(
...
         -----
-        The positions are 0-indexed.
         Operates in-place.
         """
         self._check_fit()
 
-        if mapping is not None and order is not None:
+        if (mapping is not None) + (order is not None) + (
+            template is not None
+        ) > 1:
```
Comment (mscheltienne) [inline pycrostates/cluster/_base.py:406]:
```suggestion
        if sum(x is not None for x in (mapping, order, template) > 1:
```

A bit shorter

File: pycrostates/cluster/utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    _check_type(inst, (_BaseCluster,), item_name="inst")
+    inst._check_fit()
+    _check_type(template_inst, (_BaseCluster,), item_name="template_inst")
+    template_inst._check_fit()
+
+    if inst.n_clusters != template_inst.n_clusters:
+        raise ValueError(
+            "Instance and the template must have the same "
+            "number of cluster centers"
```
Comment (mscheltienne) [inline pycrostates/cluster/utils/utils.py:50]:
```suggestion
            "Instance and the template must have the same "
            "number of cluster centers."
```

**Review by mscheltienne — APPROVED** [review]
Very minor comments. I'll just push a couple more changes.

**Comment by mscheltienne** [comment]
I added a type-hint class for Cluster, since it was not possible to type-hint to `_BaseCluster` directly (circular imports).
Alternatively, we could have used:

```
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from pycrostates.clusters._base import _BaseCluster
```

which is an actual thing in the type-hint world; and an abomination in terms of syntax.

**Comment by mscheltienne** [comment]
Looks like test are green, @vferat have a look and merge if happy.

**Pull request merged**